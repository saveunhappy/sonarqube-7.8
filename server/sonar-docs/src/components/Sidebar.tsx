/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { Link } from 'gatsby';
import CategoryBlockLink from './CategoryBlockLink';
import ExternalLink from './ExternalLink';
import PageLink from './PageLink';
import Search from './Search';
import SearchEntryResult from './SearchEntryResult';
import VersionSelect from './VersionSelect';
import DownloadIcon from './icons/DownloadIcon';
import {
  getNavTree,
  isDocsNavigationBlock,
  isDocsNavigationExternalLink,
  getOpenChainFromPath,
  testPathAgainstUrl
} from './navTreeUtils';
import { MarkdownRemark } from '../@types/graphql-types';
import { SearchResult, DocVersion, DocNavigationItem } from '../@types/types';

interface Props {
  location: Location;
  pages: MarkdownRemark[];
  version: string;
}

interface State {
  loaded: boolean;
  navTree: DocNavigationItem[];
  openChain?: DocNavigationItem[];
  query: string;
  results: SearchResult[];
  versions: DocVersion[];
}

export default class Sidebar extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    const navTree = getNavTree();
    this.state = {
      loaded: false,
      navTree,
      openChain: getOpenChainFromPath(this.props.location.pathname, navTree),
      query: '',
      results: [],
      versions: []
    };
  }

  componentDidMount() {
    this.loadVersions();
  }

  componentDidUpdate(prevProps: Props) {
    if (this.props.location.pathname !== prevProps.location.pathname) {
      this.setState(({ navTree }) => ({
        openChain: getOpenChainFromPath(this.props.location.pathname, navTree)
      }));
    }
  }

  loadVersions() {
    fetch('/DocsVersions.json')
      .then(response => response.json())
      .then(json => {
        this.setState({ loaded: true, versions: json });
      })
      .catch(() => {});
  }

  getNodeFromUrl = (url: string) => {
    return this.props.pages.find(p => {
      if (p.fields && p.fields.slug) {
        if (testPathAgainstUrl(p.fields.slug, url)) {
          return true;
        }
      }

      if (p.frontmatter && p.frontmatter.url) {
        if (testPathAgainstUrl(p.frontmatter.url, url)) {
          return true;
        }
      }

      return false;
    });
  };

  handleSearch = (results: SearchResult[], query: string) => {
    this.setState({ results, query });
  };

  renderCategory = (leaf: DocNavigationItem) => {
    if (isDocsNavigationBlock(leaf)) {
      let children: (MarkdownRemark | JSX.Element)[] = [];
      leaf.children.forEach(child => {
        if (typeof child === 'string') {
          const node = this.getNodeFromUrl(child);
          if (node) {
            children.push(node);
          }
        } else {
          children = children.concat(this.renderCategory(child));
        }
      });
      return (
        <CategoryBlockLink
          key={leaf.title}
          location={this.props.location}
          openByDefault={this.state.openChain && this.state.openChain.includes(leaf)}
          title={leaf.title}>
          {children}
        </CategoryBlockLink>
      );
    }

    if (isDocsNavigationExternalLink(leaf)) {
      return <ExternalLink external={leaf.url} key={leaf.title} title={leaf.title} />;
    }

    return (
      <PageLink
        className="page-indexes-link"
        key={leaf}
        location={this.props.location}
        node={this.getNodeFromUrl(leaf)}
      />
    );
  };

  renderCategories = () => {
    return <nav>{this.state.navTree.map(this.renderCategory)}</nav>;
  };

  renderResults = () => {
    return (
      <div>
        {this.state.results.map(result => (
          <SearchEntryResult
            active={
              (this.props.location.pathname === result.page.url && result.page.url === '/') ||
              (result.page.url !== '/' && this.props.location.pathname.endsWith(result.page.url))
            }
            key={result.page.id}
            result={result}
          />
        ))}
      </div>
    );
  };

  render() {
    const { versions } = this.state;
    const currentVersion = versions.find(v => v.current);
    const selectedVersionValue =
      currentVersion && this.props.version === 'latest' ? currentVersion.value : this.props.version;
    const isOnCurrentVersion = !currentVersion || selectedVersionValue === currentVersion.value;
    return (
      <div className="page-sidebar">
        <div className="sidebar-header">
          <Link to="/">
            <img
              alt="Continuous Code Quality"
              className="sidebar-logo"
              src={`/${this.props.version}/images/SonarQubeIcon.svg`}
              title="Continuous Code Quality"
              width="160"
            />
          </Link>
          <VersionSelect
            isOnCurrentVersion={isOnCurrentVersion}
            selectedVersionValue={selectedVersionValue}
            versions={versions}
          />
          {this.state.loaded && !isOnCurrentVersion && (
            <div className="alert alert-warning">
              This is an archived version of the doc for{' '}
              <b>SonarQube version {this.props.version}</b>. <a href="/">See Documentation</a> for
              current functionnality.
            </div>
          )}
        </div>
        <div className="page-indexes">
          <Search
            navigation={this.state.navTree}
            onResultsChange={this.handleSearch}
            pages={this.props.pages}
          />
          {this.state.query !== '' ? this.renderResults() : this.renderCategories()}
        </div>
        <div className="sidebar-footer">
          <a href="https://www.sonarqube.org/" rel="noopener noreferrer" target="_blank">
            <DownloadIcon /> SonarQube
          </a>
          <a href="https://community.sonarsource.com/" rel="noopener noreferrer" target="_blank">
            <img alt="Community" src={`/${this.props.version}/images/community.svg`} /> Community
          </a>
          <a
            className="icon-only"
            href="https://twitter.com/SonarQube"
            rel="noopener noreferrer"
            target="_blank">
            <img alt="Twitter" src={`/${this.props.version}/images/twitter.svg`} />
          </a>
          <a
            className="icon-only"
            href="https://www.sonarsource.com/resources/product-news/"
            rel="noopener noreferrer"
            target="_blank">
            <img alt="Product News" src={`/${this.props.version}/images/newspaper.svg`} />
            <span className="tooltip">Product News</span>
          </a>
        </div>
      </div>
    );
  }
}
