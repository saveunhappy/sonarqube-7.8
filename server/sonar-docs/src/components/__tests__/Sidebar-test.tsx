/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { FetchMock } from 'jest-fetch-mock';
import { shallow } from 'enzyme';
import Sidebar from '../Sidebar';
import { MarkdownRemark } from '../../@types/graphql-types';

jest.mock('../navTreeUtils', () => {
  return {
    ...require.requireActual('../navTreeUtils'),
    getNavTree: jest.fn().mockReturnValue([
      '/foo/',
      {
        title: 'Foo subs',
        children: [
          '/foo/bar/',
          '/foo/baz/',
          {
            title: 'Foo Baz subs',
            children: [
              '/foo/baz/bar/',
              '/foo/baz/foo/',
              {
                title: 'Foo Baz Foo subs',
                children: ['/foo/baz/foo/bar/', '/foo/baz/foo/baz']
              }
            ]
          }
        ]
      },
      '/bar/',
      {
        title: 'Bar subs',
        children: [{ title: 'External link 1', url: 'http://example.com/1' }, '/bar/foo/']
      },
      { title: 'External link 2', url: 'http://example.com/2' }
    ])
  };
});

beforeEach(() => {
  (fetch as FetchMock).resetMocks();
  (fetch as FetchMock).mockResponse(`[
    { "value": "2.0", "current": true },
    { "value": "1.0", "current": false }
  ]`);
});

it('should render correctly', () => {
  const wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot();
});

function shallowRender(props: Partial<Sidebar['props']> = {}) {
  return shallow(
    <Sidebar
      location={{ pathname: '/foo/baz/foo/bar' } as Location}
      pages={[
        {
          fields: {
            slug: '/foo/'
          },
          frontmatter: {
            title: 'Foo'
          }
        } as MarkdownRemark,
        {
          fields: {
            slug: '/foo/baz/bar'
          },
          frontmatter: {
            title: 'Foo Baz Bar'
          }
        } as MarkdownRemark,
        {
          fields: {
            slug: '/bar/'
          },
          frontmatter: {
            title: 'Bar'
          }
        } as MarkdownRemark
      ]}
      version="2.0"
      {...props}
    />
  );
}
