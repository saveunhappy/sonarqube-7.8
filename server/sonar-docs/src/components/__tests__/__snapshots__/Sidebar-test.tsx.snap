// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="page-sidebar"
>
  <div
    className="sidebar-header"
  >
    <ForwardRef
      to="/"
    >
      <img
        alt="Continuous Code Quality"
        className="sidebar-logo"
        src="/2.0/images/SonarQubeIcon.svg"
        title="Continuous Code Quality"
        width="160"
      />
    </ForwardRef>
    <VersionSelect
      isOnCurrentVersion={true}
      selectedVersionValue="2.0"
      versions={Array []}
    />
  </div>
  <div
    className="page-indexes"
  >
    <Search
      navigation={
        Array [
          "/foo/",
          Object {
            "children": Array [
              "/foo/bar/",
              "/foo/baz/",
              Object {
                "children": Array [
                  "/foo/baz/bar/",
                  "/foo/baz/foo/",
                  Object {
                    "children": Array [
                      "/foo/baz/foo/bar/",
                      "/foo/baz/foo/baz",
                    ],
                    "title": "Foo Baz Foo subs",
                  },
                ],
                "title": "Foo Baz subs",
              },
            ],
            "title": "Foo subs",
          },
          "/bar/",
          Object {
            "children": Array [
              Object {
                "title": "External link 1",
                "url": "http://example.com/1",
              },
              "/bar/foo/",
            ],
            "title": "Bar subs",
          },
          Object {
            "title": "External link 2",
            "url": "http://example.com/2",
          },
        ]
      }
      onResultsChange={[Function]}
      pages={
        Array [
          Object {
            "fields": Object {
              "slug": "/foo/",
            },
            "frontmatter": Object {
              "title": "Foo",
            },
          },
          Object {
            "fields": Object {
              "slug": "/foo/baz/bar",
            },
            "frontmatter": Object {
              "title": "Foo Baz Bar",
            },
          },
          Object {
            "fields": Object {
              "slug": "/bar/",
            },
            "frontmatter": Object {
              "title": "Bar",
            },
          },
        ]
      }
    />
    <nav>
      <PageLink
        className="page-indexes-link"
        key="/foo/"
        location={
          Object {
            "pathname": "/foo/baz/foo/bar",
          }
        }
        node={
          Object {
            "fields": Object {
              "slug": "/foo/",
            },
            "frontmatter": Object {
              "title": "Foo",
            },
          }
        }
      />
      <CategoryLink
        key="Foo subs"
        location={
          Object {
            "pathname": "/foo/baz/foo/bar",
          }
        }
        openByDefault={true}
        title="Foo subs"
      >
        <CategoryLink
          key="Foo Baz subs"
          location={
            Object {
              "pathname": "/foo/baz/foo/bar",
            }
          }
          openByDefault={true}
          title="Foo Baz subs"
        >
          <Component />
          <CategoryLink
            key="Foo Baz Foo subs"
            location={
              Object {
                "pathname": "/foo/baz/foo/bar",
              }
            }
            openByDefault={true}
            title="Foo Baz Foo subs"
          />
        </CategoryLink>
      </CategoryLink>
      <PageLink
        className="page-indexes-link"
        key="/bar/"
        location={
          Object {
            "pathname": "/foo/baz/foo/bar",
          }
        }
        node={
          Object {
            "fields": Object {
              "slug": "/bar/",
            },
            "frontmatter": Object {
              "title": "Bar",
            },
          }
        }
      />
      <CategoryLink
        key="Bar subs"
        location={
          Object {
            "pathname": "/foo/baz/foo/bar",
          }
        }
        openByDefault={false}
        title="Bar subs"
      >
        <ExternalLink
          external="http://example.com/1"
          key="External link 1"
          title="External link 1"
        />
      </CategoryLink>
      <ExternalLink
        external="http://example.com/2"
        key="External link 2"
        title="External link 2"
      />
    </nav>
  </div>
  <div
    className="sidebar-footer"
  >
    <a
      href="https://www.sonarqube.org/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <DownloadIcon />
       SonarQube
    </a>
    <a
      href="https://community.sonarsource.com/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        alt="Community"
        src="/2.0/images/community.svg"
      />
       Community
    </a>
    <a
      className="icon-only"
      href="https://twitter.com/SonarQube"
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        alt="Twitter"
        src="/2.0/images/twitter.svg"
      />
    </a>
    <a
      className="icon-only"
      href="https://www.sonarsource.com/resources/product-news/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        alt="Product News"
        src="/2.0/images/newspaper.svg"
      />
      <span
        className="tooltip"
      >
        Product News
      </span>
    </a>
  </div>
</div>
`;
